src/event/vrl_target.rs

```diff
@@ -0,0 +1,395 @@
+use super::{Event, EventMetadata, LogEvent, Metric, MetricKind, Value};
+use crate::config::log_schema;
+use snafu::Snafu;
+use std::{collections::BTreeMap, convert::TryFrom, iter::FromIterator};
+use vrl::path::Segment;
+
+const VALID_METRIC_PATHS_SET: &str = ".name, .namespace, .timestamp, .kind, .tags";
+
+/// We can get the `type` of the metric in Remap, but can't set  it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// An adapter to turn `Event`s into `vr::Target`s
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
+    LogEvent(vrl::Value, EventMetadata),
+    Metric(Metric),
+}
+
+impl VrlTarget {
+    pub fn new(event: Event) -> Self {
+        match event {
+            Event::Log(event) => {
+                let metadata = event.metadata().to_owned();
+                let fields: BTreeMap<String, Value> = event.into();
+                let value: Value = fields.into();
+                VrlTarget::LogEvent(value.into(), metadata)
+            }
+            Event::Metric(event) => VrlTarget::Metric(event),
+        }
+    }
+
+    /// Turn the target back into events
+    ///
+    /// This returns an iterator of events as one event can be turned into multiple by assign `.`
+    /// to an array in VRL
```

Comment:
```suggestion
    /// Turn the target back into events.
    ///
    /// This returns an iterator of events as one event can be turned into multiple events by
    /// assigning an array to `.` in VRL.
```

Reply:
👍 we seem to have a mix of doc comments with and without `.`s. I wonder if there is a way we could enforce `.`s if we prefer that. I don't have a strong opinion.